plutus-ledger-api/src/Plutus/V1/Ledger/Scripts.hs

```diff
@@ -181,13 +183,13 @@ evaluateScript s = do
     p <- case PLC.runQuote $ runExceptT @PLC.FreeVariableError $ UPLC.unDeBruijnProgram namedProgram of
         Right p -> return p
         Left e  -> throwError $ MalformedScript $ Haskell.show e
-    let (logOut, _tally, result) = evaluateCekTrace p
+    let (logOut, UPLC.TallyingSt _ budget, result) = evaluateCekTrace p
```

Comment:
`evaluateScriptCounting` uses a slightly different way to run the script and get the budget. It also takes a serialised script + arguments, which we don't have/need here. Maybe there's a way to generalise `evaluateScriptCounting` so that we can call it from here (I'd just prefer to use exactly the same way of computing the cost here)

Reply:
Isn't it really wasteful to do evaluation twice? `evaluateScriptCounting` also just folds the arguments with `apply`, it doesn't seem that different. But this is an opportunity to add `evaluateScriptCounting`, and compare the results and maybe adapt `evaluateScript` so that the results become the same?

Reply:
Can't we use `evaluateScriptRestricting`, to really match up? `evaluateScriptRestricting` has the leftover budget and could return it, in which case you can take the difference between the budget provided and the budget left to see how much was used.

Reply:
But `evaluateScriptRestricting` needs an `ExBudget` which we would first need to calculate, so that would still mean 2 evaluations?

Reply:
Oh right, I forgot we don't actually have user-specified budgets on transactions yet.

Reply:
> do evaluation twice

I was thinking we could completely replace `evaluateScript` with `evaluateScriptCounting` in the emulator. I just wanted to minimise the divergence between the two functions.

Reply:
I have refactored the code so that both `evaluateScript` and `evaluateScriptCounting` use the same term-building code. The error handling is a bit different so merging them didn't really work.